### PR TITLE
Response header annotation

### PR DESCRIPTION
**Pull request by pristupa**

### CODE REVIEW

File: winter/http/response_header.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        self._headers = headers
+        self._header_name = header_name
+
+    def set(self, value: T):
+        self._headers[self._header_name] = convert(value, self._value_type)
+
+
+class _ResponseHeader:
+    def __getitem__(self, value_type: Type):
+        assert isinstance(value_type, type), 'value_type must be a type'
```
Comment (berenda-andrey) [inline winter/http/response_header.py]: isinstance(typing.Dict, type) is False
Is it good?

File: winter/http/response_header_resolver.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+from .response_header import ResponseHeaderAnnotation
+from .response_header import ResponseHeaders
+from .. import ArgumentResolver
+from ..core import ComponentMethodArgument
+
+
+class ResponseHeaderArgumentResolver(ArgumentResolver):
+
+    def is_supported(self, argument: ComponentMethodArgument) -> bool:
+        if not isinstance(argument.type_, type) or not issubclass(argument.type_, _BaseResponseHeader):
```
Comment (berenda-andrey) [inline winter/http/response_header_resolver.py]: the same as above: isinstance(typing.Dict, type) is False

File: winter/http/response_header.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        payload = {
+            '_value_type': value_type,
+        }
+        return type(f'ResponseHeader[{value_type.__name__}]', (_BaseResponseHeader,), payload)
```
Comment (berenda-andrey) [inline winter/http/response_header.py]: typing.Dict.__name__  raises AttributeError: __name__

File: tests/controllers/controller_with_response_headers.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+import winter
+from winter import ResponseHeader
+
+
+@winter.controller
+@winter.route('with-response-headers/')
+class ControllerWithResponseHeaders:
+
+    @winter.response_header('x-header', 'header')
```
Comment (berenda-andrey) [inline tests/controllers/controller_with_response_headers.py]: I suggest to move to @winter.http.response_header(...)

Comment (pristupa) [reply tests/controllers/controller_with_response_headers.py]: What about existing winter.request_body then?

Comment (berenda-andrey) [reply tests/controllers/controller_with_response_headers.py]: seems wrong

File: winter/http/response_header.py (hunk 1 above)
Comment (pristupa) [reply winter/http/response_header.py]: not quite good, any ideas?

File: winter/http/response_header.py (hunk 3 above)
Comment (pristupa) [reply winter/http/response_header.py]: should we leave ResponseHeader[] for example?

File: tests/controllers/controller_with_response_headers.py (hunk 4 above)
Comment (pristupa) [reply tests/controllers/controller_with_response_headers.py]: so I remove both request_body and request_header from init.py, leaving them only in http's init.py?

Comment (pristupa) [reply tests/controllers/controller_with_response_headers.py]: what to do with looking similar response_status (now it's not even in http directory)?

File: winter/http/response_header_resolver.py (hunk 2 above)
Comment (pristupa) [reply winter/http/response_header_resolver.py]: fixed

File: winter/schema/generation.py [hunk 5]
```diff
@@ -15,8 +15,8 @@
 from ..drf import get_output_serializer
 from ..exceptions.handlers import MethodExceptionsHandler
 from ..exceptions.handlers import exceptions_handler
-from ..response_entity import ResponseEntity
-from ..response_status import get_default_response_status
+from winter.http.response_entity import ResponseEntity
```
Comment (berenda-andrey) [inline winter/schema/generation.py]: it is better to use relative import

File: winter/routing/argument_resolvers/path_parameters_argument_resolver.py [hunk 6]
```diff
@@ -1,6 +1,7 @@
 from django.urls import get_resolver
 from rest_framework.request import Request
 
+from winter.http import ResponseHeaders
```
Comment (berenda-andrey) [inline winter/routing/argument_resolvers/path_parameters_argument_resolver.py]: it is better to use relative import

File: winter/http/response_status.py [hunk 7]
```diff
@@ -3,8 +3,8 @@
 
 import dataclasses
 
-from .core import ComponentMethod
-from .core import annotate
+from winter.core import ComponentMethod
```
Comment (berenda-andrey) [inline winter/http/response_status.py]: it is better to use relative import

File: winter/http/response_header_resolver.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
+from rest_framework.request import Request
+
+from winter.http.response_header import _BaseResponseHeader
```
Comment (berenda-andrey) [inline winter/http/response_header_resolver.py]: it is better to use relative import

File: winter/http/response_header.py (hunk 3 above)
Comment (berenda-andrey) [reply winter/http/response_header.py]: I suggest f'ResponseHeader[{value_type}]'

File: winter/http/response_header.py (hunk 1 above)
Comment (berenda-andrey) [reply winter/http/response_header.py]: I was wrong isinstance(typing.Dict, type) is True

Comment (berenda-andrey) [reply winter/http/response_header.py]: seems only typing.Any not isinstance type, typing.Dict, typing.Dict[str, str] or something like that are instance type

File: winter/http/response_header.py (hunk 3 above)
Comment (berenda-andrey) [reply winter/http/response_header.py]: typing.Dict[str, str].__name__ is Dict, and str(typing.Dict[str, str].__name__) is typing.Dict[str, str]

File: tests/controllers/controller_with_response_headers.py (hunk 4 above)
Comment (berenda-andrey) [reply tests/controllers/controller_with_response_headers.py]: Also we can move it in http(may be in different pr may be in this pr)

File: winter/http/response_header_resolver.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+from .response_header import ResponseHeaderAnnotation
+from .response_header import ResponseHeaders
+from .. import ArgumentResolver
+from ..core import ComponentMethodArgument
+
+
+class ResponseHeaderArgumentResolver(ArgumentResolver):
+
+    def is_supported(self, argument: ComponentMethodArgument) -> bool:
+        if not is_origin_type_subclasses(argument.type_, _BaseResponseHeader):
```
Comment (berenda-andrey) [inline winter/http/response_header_resolver.py]: it seems just inspect.isclass(argument.type_) and issubclass(argument.type_, check_class) - enough

File: winter/__init__.py [hunk 10]
```diff
@@ -10,15 +10,15 @@
 from .drf import output_template
 from .exceptions import RedirectException
 from .exceptions.handlers import ExceptionHandler
-from .exceptions.handlers import RedirectExceptionHandler
+from winter.http.exception_handlers import RedirectExceptionHandler
```
Comment (berenda-andrey) [inline winter/__init__.py]: relative import

File: winter/__init__.py [hunk 11]
```diff
@@ -29,32 +29,56 @@
...
+
+
+def _register_output_processor_resolvers():
+    from .pagination.page import PageOutputProcessorResolver
+
     register_output_processor_resolver(PageOutputProcessorResolver())
+
+
+def _add_exception_handlers():
+    from winter.http.exception_handlers import ConvertExceptionHandler
```
Comment (berenda-andrey) [inline winter/__init__.py]: relative import

File: winter/http/default_response_status.py [hunk 12]
```diff
@@ -1,10 +1,7 @@
 from http import HTTPStatus
-from typing import Union
 
-import dataclasses
-
-from .core import ComponentMethod
-from .core import annotate
+from winter.core import ComponentMethod
```
Comment (berenda-andrey) [inline winter/http/default_response_status.py]: relative import

File: winter/__init__.py (hunk 10 above)
Comment (mofr) [inline winter/__init__.py]: Please use relative import

File: winter/http/default_response_status.py [hunk 13]
```diff
@@ -1,10 +1,7 @@
...
+from winter.http.response_status_annotation import ResponseStatusAnnotation
```
Comment (mofr) [inline winter/http/default_response_status.py]: Please use relative imports

File: winter/http/response_header_annotation.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+import dataclasses
+
+from ..converters import convert
+from ..core import annotate_method
+
+T = TypeVar('T')
+
+
+class _BaseResponseHeader:
```
Comment (berenda-andrey) [inline winter/http/response_header_annotation.py]:
```
class _ResponseHeaderMeta(type):
    def __getitem__(self, value_type):
        assert isinstance(value_type, type), 'value_type must be a type'
        return _ResponseHeaderMeta.__new__(
            type(self),
            f'{self.__name__}[{value_type.__name__}]',
            (self, ),
            {
                '_value_type': value_type,
            },

        )


class ResponseHeader(metaclass=_ResponseHeaderMeta):
    _value_type = str

    def __init__(self, headers, header_name):
        self._headers = headers
        self._header_name = header_name

    def __str__(self):
        return f'{self.__class__.__name__}(headers={self._headers}, header_name={self._header_name})'

    def __repr__(self):
        return str(self)

    def set(self, value):
        self._headers[self._header_name] = value
```
Кажется так немного красивее получится

Comment (berenda-andrey) [reply winter/http/response_header_annotation.py]: и не будет глабальных переменных

Comment (berenda-andrey) [reply winter/http/response_header_annotation.py]: и можно будет проверить так
issubclass(ResponseHeader[int], ResponseHeader)

Comment (pristupa) [reply winter/http/response_header_annotation.py]: Я стырил реализацию с Generic-сериалайзеров. Будет круто, если запилишь сразу в два места :)

Comment (pristupa) [reply winter/http/response_header_annotation.py]: запилил

Comment (pristupa) [reply winter/http/response_header_annotation.py]: спасибо за отличную идею

**Pull request merged**